#[implement(IBufferByteAccess)] gives this `struct takes 4 arguments but 7 arguments were supplied`

I'm trying to do a Rust version of [this](https://gist.github.com/kennykerr/39752b796e774957f4d55d1b4edd79cc):
```rust
#[implement(Windows::Win32::System::WinRT::IBufferByteAccess)]
pub struct WBuffer {
    bytes: &'static mut [u8],
    length: u32
}

impl WBuffer { ... }
```

However I get this error:
```error[E0061]: this struct takes 4 arguments but 7 arguments were supplied
   --> src\engine\bindings\src\custom.rs:5:1
    |
5   |   #[implement(Windows::Win32::System::WinRT::IBufferByteAccess)]
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   |
    |   supplied 7 arguments
    |   expected 4 arguments
    |   in this procedural macro expansion
    | 
   ::: C:\Users\enigm\.cargo\registry\src\github.com-1ecc6299db9ec823\windows_macros-0.17.2\src\lib.rs:164:1
    |
164 | / pub fn implement(
165 | |     attribute: proc_macro::TokenStream,
166 | |     input: proc_macro::TokenStream,
167 | | ) -> proc_macro::TokenStream {
    | |____________________________- in this expansion of `#[implement]`

```

Looking through the generated code (using `cargo-expand`), there's one place where the error looks like it occurs:
```rust
    impl WBuffer_box {
        const VTABLES: (Windows::Win32::System::WinRT::IBufferByteAccess_abi,) =
            (Windows::Win32::System::WinRT::IBufferByteAccess_abi(
                Self::QueryInterface_abi0,
                Self::AddRef_abi0,
                Self::Release_abi0,
                Self::GetIids,
                Self::GetRuntimeClassName,
                Self::GetTrustLevel,
                Self::abi0_6,
            ),);
        const IDENTITY_VTABLE: ::windows::IInspectable_abi = ::windows::IInspectable_abi(
            Self::identity_query_interface,
            Self::identity_add_ref,
            Self::identity_release,
            Self::GetIids,
            Self::GetRuntimeClassName,
            Self::GetTrustLevel,
        );
...
```
I'm not quite sure why `IBufferByteAccess_abi` has only 4 arguments, though. Or why 7 were provided as above.

--- a/crates/gen/src/parser/method_signature.rs
+++ b/crates/gen/src/parser/method_signature.rs
@@ -5,6 +5,7 @@ pub struct MethodSignature {
     pub return_type: Option<Signature>,
 }
 
+#[derive(Clone)]
 pub struct MethodParam {
     pub param: tables::Param,
     pub signature: Signature,
@@ -449,7 +450,7 @@ impl MethodSignature {
 impl MethodParam {
     fn is_retval(&self) -> bool {
         let flags = self.param.flags();
-        if flags.input() || !flags.output() || self.signature.pointers != 1 {
+        if flags.input() || !flags.output() {
             return false;
         }
 

--- a/crates/gen/src/types/com_interface.rs
+++ b/crates/gen/src/types/com_interface.rs
@@ -207,10 +207,10 @@ fn gen_method(
         let args = leading_params.iter().map(|p| p.gen_win32_abi_arg());
 
         let mut return_param = signature.params[signature.params.len() - 1].clone();
-        
+
         let return_type_tokens = if return_param.signature.pointers > 1 {
             return_param.signature.pointers -= 1;
-             return_param.gen_win32(gen) 
+            return_param.gen_win32(gen)
         } else {
             return_param.signature.kind.gen_name(gen)
         };

--- a/crates/gen/src/types/com_interface.rs
+++ b/crates/gen/src/types/com_interface.rs
@@ -206,13 +206,14 @@ fn gen_method(
         let params = signature.gen_win32_params(leading_params, gen);
         let args = leading_params.iter().map(|p| p.gen_win32_abi_arg());
 
-        let return_type_tokens = signature
-            .params
-            .last()
-            .unwrap()
-            .signature
-            .kind
-            .gen_name(gen);
+        let mut return_param = signature.params[signature.params.len() - 1].clone();
+        
+        let return_type_tokens = if return_param.signature.pointers > 1 {
+            return_param.signature.pointers -= 1;
+             return_param.gen_win32(gen) 
+        } else {
+            return_param.signature.kind.gen_name(gen)
+        };
 
         quote! {
             pub unsafe fn #name<#constraints>(&self, #params) -> ::windows::Result<#return_type_tokens> {

--- a/src/traits/abi.rs
+++ b/src/traits/abi.rs
@@ -47,6 +47,11 @@ pub unsafe trait Abi: Sized + Clone {
     fn drop_param(_: &mut Param<Self>) {}
 }
 
+unsafe impl<T> Abi for *mut T {
+    type Abi = Self;
+    type DefaultType = Self;
+}
+
 unsafe impl<T: Interface> Abi for T {
     type Abi = RawPtr;
     type DefaultType = Option<T>;

--- a/tests/implement/build.rs
+++ b/tests/implement/build.rs
@@ -2,8 +2,11 @@ fn main() {
     windows::build! {
         Windows::Foundation::Collections::{IIterable, IVectorView},
         Windows::Foundation::{IClosable, IStringable, Uri},
+        Windows::Storage::Streams::Buffer,
         Windows::Win32::Foundation::E_BOUNDS,
-        Windows::Win32::System::WinRT::{IDisplayPathInterop, ISwapChainInterop},
+        Windows::Win32::System::WinRT::{
+            IBufferByteAccess, IDisplayPathInterop, ISwapChainInterop,
+        },
         Windows::UI::Xaml::{Application, Controls::Button},
     };
 }

--- a/tests/implement/tests/buffer.rs
+++ b/tests/implement/tests/buffer.rs
@@ -1,0 +1,36 @@
+use test_implement::*;
+use windows::*;
+
+use Windows::Storage::Streams::Buffer;
+use Windows::Win32::System::WinRT::IBufferByteAccess;
+
+#[implement(Windows::Win32::System::WinRT::IBufferByteAccess)]
+struct TestBuffer(Vec<u8>);
+
+#[allow(non_snake_case)]
+impl TestBuffer {
+    fn Buffer(&mut self) -> Result<*mut u8> {
+        Ok(self.0.as_mut_ptr())
+    }
+}
+
+#[test]
+fn test() -> Result<()> {
+    let object: IBufferByteAccess = TestBuffer(vec![0xAA, 0xBB, 0xCC]).into();
+
+    let bytes: *const u8 = unsafe { object.Buffer()? };
+    let bytes = unsafe { std::slice::from_raw_parts(bytes, 3) };
+    assert_eq!(bytes, [0xAA, 0xBB, 0xCC]);
+
+    let object = Buffer::Create(3)?;
+
+    let bytes: *mut u8 = unsafe { object.cast::<IBufferByteAccess>()?.Buffer()? };
+    let bytes = unsafe { std::slice::from_raw_parts_mut(bytes, 3) };
+    bytes.copy_from_slice(&[0xCC, 0xBB, 0xAA]);
+
+    let bytes: *const u8 = unsafe { object.cast::<IBufferByteAccess>()?.Buffer()? };
+    let bytes = unsafe { std::slice::from_raw_parts(bytes, 3) };
+
+    assert_eq!(bytes, [0xCC, 0xBB, 0xAA]);
+    Ok(())
+}

